### PR TITLE
ESQL: Extend STATS command to support aggregate expressions

### PR DESCRIPTION
**Pull request by costin**
Previously only aggregate functions (max/sum/etc..) were allowed inside
 the stats command. This PR allows expressions involving one or multiple
 aggregates to be used, such as:
```
 stats x = avg(salary % 3) + max(emp_no),
       y = min(emp_no / 3) + 10 - median(salary)
       by z = languages % 2
```

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:1]: The core of this PR - the changes are larger because:
1. the ReplaceDuplicateAggWithEval has been removed and incorporated into the new rule.
2. the behavior of CombineProjection has been fixed when dealing with a Project/Aggregate, simplifying the clean-up.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 1]
```diff
@@ -239,15 +239,18 @@ protected LogicalPlan rule(Aggregate aggregate) {
                     // project away transient fields and re-enforce the original order using references (not copies) to the original aggs
                     // this works since the replaced aliases have their nameId copied to avoid having to update all references (which has
                     // a cascading effect)
-                    plan = new EsqlProject(source, plan, Expressions.asAttributes(aggs));
+                    plan = new Project(source, plan, Expressions.asAttributes(aggs));
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:242]: Small tweak - no need to return EsqlProject instead Project since the tree is already resolved.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 2]
```diff
@@ -239,15 +239,18 @@ protected LogicalPlan rule(Aggregate aggregate) {
...
                 }
             }
 
             return plan;
         }
 
-        static String temporaryName(NamedExpression agg, AggregateFunction af) {
-            return "__" + agg.name() + "_" + af.functionName() + "@" + Integer.toHexString(af.hashCode());
+        static String temporaryName(Expression expression, AggregateFunction af) {
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:249]: I've updated the temporary name to make it a bit less confusing/repetitive.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 3]
```diff
@@ -259,17 +262,23 @@ static class ConvertStringToByteRef extends OptimizerRules.OptimizerExpressionRu
...
                 }
-                return Literal.of(lit, l.stream().map(v -> new BytesRef((String) v)).toList());
+                List<BytesRef> byteRefs = new ArrayList<>(l.size());
+                for (Object v : l) {
+                    byteRefs.add(new BytesRef(v.toString()));
+                }
+                return Literal.of(lit, byteRefs);
             }
             return lit;
         }
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:284]: Unrelated change - couldn't help not correct it to save the method invocation and replace the noise map with a good-ol' reliable iteration.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 4]
```diff
@@ -288,39 +297,86 @@ protected LogicalPlan rule(UnaryPlan plan) {
...
-                    var newAggs = combineProjections(project.projections(), aggs);
-                    var newGroups = replacePrunedAliasesUsedInGroupBy(a.groupings(), aggs, newAggs);
-                    return new Aggregate(a.source(), a.child(), newGroups, newAggs);
+                    var tuple = projectAggregations(project.projections(), aggs);
+                    // project can be fully removed
+                    if (tuple.v1().isEmpty()) {
+                        var newAggs = tuple.v2();
+                        var newGroups = replacePrunedAliasesUsedInGroupBy(a.groupings(), aggs, newAggs);
+                        plan = new Aggregate(a.source(), a.child(), newGroups, newAggs);
+                    }
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:317]: The gist of this change takes care of the scenario creates by  `ReplaceStatsAggExpressionWithEval`:
```
stats x = sum(), y = count()
project x, x as a, y
```

The combine rule previously would combine project into stats which would duplicate the count:
```
stats x = sum(), a = sum(), y = count()
```
It removes the project (which is cheap) but duplicates the sum (which is expensive and the reason we didn't want to duplicate it in the first place).

The rule thus tracks is there's any new alias - however it keeps on removing unused aggregations and in case of  basic aliasing project, removes the project.
So the following
```
stats x = sum(), y = count()
project x as a
```

becomes
```
stats a = sum()
```

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 5]
```diff
@@ -366,10 +422,7 @@ private List<Expression> replacePrunedAliasesUsedInGroupBy(
         }
 
         public static Expression trimNonTopLevelAliases(Expression e) {
-            if (e instanceof Alias a) {
-                return new Alias(a.source(), a.name(), a.qualifier(), trimAliases(a.child()), a.id());
-            }
-            return trimAliases(e);
+            return e instanceof Alias a ? a.replaceChild(trimAliases(a.child())) : trimAliases(e);
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:425]: Ternary operator ❤️

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 6]
```diff
@@ -1135,7 +1183,7 @@ protected LogicalPlan rule(Aggregate aggregate) {
                         if (field instanceof Attribute == false && field.foldable() == false) {
                             // 3. create a new alias if one doesn't exist yet no reference
                             Attribute attr = expToAttribute.computeIfAbsent(field.canonical(), k -> {
-                                Alias newAlias = new Alias(k.source(), temporaryName(agg, af), null, k, null, true);
+                                Alias newAlias = new Alias(k.source(), temporaryName(k, af), null, k, null, true);
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:1186]: Update the generator name strategy to be a bit more meaningful.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 7]
```diff
@@ -1165,8 +1213,134 @@ protected LogicalPlan rule(Aggregate aggregate) {
...
+     * stats a = sum(a) + min(b) by x
+     * becomes
+     * stats a1 = sum(a), a2 = min(b) by x | eval a = a1 + a2 | keep a, x
+     *
+     * Since the logic is very similar, this rule also handles duplicate aggregate functions to avoid duplicate compute
+     * stats a = min(x), b = min(x), c = count(*), d = count() by g
+     * becomes
+     * stats a = min(x), c = count(*) by g | eval b = a, d = c | keep a, b, c, d, g
+     */
+    static class ReplaceStatsAggExpressionWithEval extends OptimizerRules.OptimizerRule<Aggregate> {
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:1232]: The core of this PR - breaks down the expression over aggs and adds an eval lazily only for the fields that have an expression over aggregate functions.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec [hunk 8]
```diff
@@ -1103,3 +1103,37 @@ STATS ck = COUNT(job_positions),
 ck:l | cb:l | cd:l | ci:l | c:l | csv:l 
 221  | 204  | 183  | 183  | 100 | 100
 ;
+
+nestedAggsNoGrouping
+FROM employees
+| STATS x = AVG(salary) /2 + MAX(salary), a = AVG(salary), m = MAX(salary)
```
Comment (costin) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:1109]: /cc @leemthompo

Comment (leemthompo) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:1118]: @costin I understand that we tag some of these tests to be included as examples in the docs. Just wondering what the workflow was with Abdon to add these tags? Was it just simply a ping to alert the writer that we want this example in the docs? :)

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java [hunk 9]
```diff
@@ -2895,6 +2921,216 @@ public void testNestedMultiExpressionsInGroupingAndAggs() {
...
+
+        var project = as(plan, Project.class);
+        assertThat(Expressions.names(project.projections()), contains("x"));
+        var eval = as(project.child(), Eval.class);
+        var fields = eval.fields();
+        assertThat(Expressions.name(fields.get(1)), is("x"));
+        // sum/count to compute avg
+        var div = as(fields.get(0).child(), Div.class);
+        // avg + max
+        var add = as(fields.get(1).child(), Add.class);
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:2948]: We're technically not asserting the whole expression here - maybe better to just assert the expression string?

Applies to all added tests in this file.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java [hunk 10]
```diff
@@ -2895,6 +2921,216 @@ public void testNestedMultiExpressionsInGroupingAndAggs() {
...
+     *        languages{f}#1926 % 2[INTEGER] AS z]]
+     *   \_Limit[500[INTEGER]]
+     *     \_Aggregate[[z{r}#1922],[SUM(____x_AVG@e03a7a5c_AVG@e03a7a5c{r}#1937) AS ____x_AVG@e03a7a5c_SUM@acf776e3,
+     *     COUNT(____x_AVG@e03a7a5c_AVG@e03a7a5c{r}#1937) AS ____x_AVG@e03a7a5c_COUNT@39975522,
+     *     MAX(emp_no{f}#1923) AS __x_MAX@4e09cbd5,
+     *     MIN(____y_MIN@80cee21c_MIN@80cee21c{r}#1938) AS __y_MIN@80cee21c,
+     *     PERCENTILE(salary{f}#1928,50[INTEGER]) AS __y_MEDIAN@705fccec]]
+     *       \_Eval[[languages{f}#1926 % 2[INTEGER] AS z,
+     *               salary{f}#1928 % 3[INTEGER] AS ____x_AVG@e03a7a5c_AVG@e03a7a5c,
+     *               emp_no{f}#1923 / 3[INTEGER] AS ____y_MIN@80cee21c_MIN@80cee21c]]
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:2976]: nit: the attribute names are constructed from the aggs where they will be used, but not from the expression where they will be used; this makes it a bit hard to read this bottom-up (and top-down is also hard). I think it will also lead to tricky attribute names in cases like `y = min(emp_no /3) - min(emp_no + 2)`, as both attributes will be called `____y_MIN@...._MIN@....`.
Maybe this would be more consistent?
```suggestion
     *               emp_no{f}#1923 / 3[INTEGER] AS y_SUB1_MIN]]
```
(resp. `...SUB0...` for the left hand side)

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 11]
```diff
@@ -1165,8 +1213,134 @@ protected LogicalPlan rule(Aggregate aggregate) {
...
+            Holder<Boolean> changed = new Holder<>(false);
+
+            for (NamedExpression agg : aggs) {
+                if (agg instanceof Alias as) {
+                    // if the child a nested expression
+                    Expression child = as.child();
+
+                    // common case - handle duplicates
+                    if (child instanceof AggregateFunction af) {
+                        AggregateFunction canonical = (AggregateFunction) af.canonical();
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:1265]: Interesting. Wondering if it's worth double checking this in a test of the form `stats avg(a+b), avg(b+a)` (using the canonical form of commutative operators).

**Review by astefan — APPROVED** [review]
LGTM
Left only minor comments. Maybe the one related to an additional test to be of more importance.

**Review by alex-spies — COMMENTED** [review]
Gave this a first round, focusing on tests this time. I'll give this another go tomorrow.

Two observations:
- I think there's bugs for the following cases: `stats max(l) by l=languages` (verification exception) and, more severely, `stats max(languages) + languages by l = languages` (NPE, `Cannot invoke \"org.elasticsearch.xpack.esql.planner.Layout$ChannelAndType.channel()\" because the return value of \"org.elasticsearch.xpack.esql.planner.Layout.get(org.elasticsearch.xpack.ql.expression.NameId)\" is null"`); the latter works fine without the alias.
- This allows shenanigans like `stats max(languages) + languages by languages` (using a grouping in the expression), but not `stats languages + 1 by languages`; although that may be something for a follow-up, if we want to allow this.

Other than that I have mostly minor remarks; the tests do not fully assert the (complex) expressions that are being constructed, maybe we should be stricter there.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec (hunk 8 above)
Comment (costin) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:1118]: The tests above are meant for internal consumption hence the ping - better to create other tests, that fix the general dataset and the rest of examples in the docs and add them in, as a separate PR after this one gets merged.
See the previous PRs authored by Abdon.

**Review by fang-xing-esql — COMMENTED** [review]
LGTM in general, this is more like educational purposes for me. I left two minor suggestions on the test cases, and a question on the CombineProjections rule.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java (hunk 7 above)
Comment (bpintea) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:1232]: We could improve this a bit further for a referencing case: just like we now support `| eval x = field, y = x + 1`, we could now (but don't yet) support something like: `| stats x = max(field), y = x + min(field)` -- this now fails because column `x` isn't known.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java [hunk 12]
```diff
@@ -2895,6 +2921,216 @@ public void testNestedMultiExpressionsInGroupingAndAggs() {
...
+     * Project[[a{r}#5, b{r}#9, __b_COUNT@a31a3f25{r}#46 AS d, __c_MIN@33a7e8da{r}#47 AS e, __a_MAX@3564a755{r}#45 AS g]]
+     * \_Eval[[____a_AVG@ff859ba4_SUM@c0c1a11f{r}#48 / __b_COUNT@a31a3f25{r}#46 AS __a_AVG@ff859ba4, __a_AVG@ff859ba4{r}#44
+     * + __a_MAX@3564a755{r}#45 AS a, __a_MAX@3564a755{r}#45 + 3[INTEGER] + 3.141592653589793[DOUBLE] + __b_COUNT@a31a3f25{r}#46 AS b]]
+     *   \_Limit[500[INTEGER]]
+     *     \_Aggregate[[w{r}#28],[SUM(salary{f}#39) AS ____a_AVG@ff859ba4_SUM@c0c1a11f, MAX(salary{f}#39) AS __a_MAX@3564a755, COUNT(s
+     * alary{f}#39) AS __b_COUNT@a31a3f25, MIN(salary{f}#39) AS __c_MIN@33a7e8da]]
+     *       \_Eval[[languages{f}#37 % 2[INTEGER] AS w]]
+     *         \_EsRelation[test][_meta_field{f}#40, emp_no{f}#34, first_name{f}#35, ..]
+     */
+    public void testStatsExpOverAggsWithScalarAndDuplicateAggs() {
```
[the remark this answers is not in the document]
Comment (alex-spies) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:3092]: I think this would be covered as part of our plan to eliminate common (sub-) expressions: https://github.com/elastic/elasticsearch/issues/103301

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 13]
```diff
@@ -1165,8 +1213,134 @@ protected LogicalPlan rule(Aggregate aggregate) {
...
+    /**
+     * Replace nested expressions over aggregates with synthetic eval post the aggregation
+     * stats a = sum(a) + min(b) by x
+     * becomes
+     * stats a1 = sum(a), a2 = min(b) by x | eval a = a1 + a2 | keep a, x
+     *
+     * Since the logic is very similar, this rule also handles duplicate aggregate functions to avoid duplicate compute
+     * stats a = min(x), b = min(x), c = count(*), d = count() by g
+     * becomes
+     * stats a = min(x), c = count(*) by g | eval b = a, d = c | keep a, b, c, d, g
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:1230]: ++ to the examples in the javadoc, very useful.

**Review by alex-spies — COMMENTED** [review]
Gave it another round, focusing on the optimizer code this time. I didn't find anything not already covered by others' remarks. Looks good, very neat feature!

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 14]
```diff
@@ -288,39 +297,86 @@ protected LogicalPlan rule(UnaryPlan plan) {
...
 
+        // variant of #combineProjections specialized for project followed by agg due to the rewrite rules applied on aggregations
+        // this method tries to combine the projections by paying attention to:
+        // - aggregations that are projected away - remove them
+        // - aliases in the project that point to aggregates - keep them in place (to avoid duplicating the aggs)
+        private Tuple<List<? extends NamedExpression>, List<? extends NamedExpression>> projectAggregations(
+            List<? extends NamedExpression> upperProjection,
+            List<? extends NamedExpression> lowerAggregations
+        ) {
+            AttributeMap<Expression> lowerAliases = new AttributeMap<>();
```
Comment (bpintea) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:340]: `lowerAliases` aren't used.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 15]
```diff
@@ -288,39 +297,86 @@ protected LogicalPlan rule(UnaryPlan plan) {
...
+            }
+
+            // remove lower aggs that are not referenced in the upper projections
+            List<? extends NamedExpression> keptProjections = upperProjection;
+            if (keepProject == false) {
+                keptProjections = emptyList();
+                lowerAggregations = combineProjections(upperProjection, lowerAggregations);
+            }
+
+            return new Tuple<>(keptProjections, lowerAggregations);
```
Comment (bpintea) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:365]: the Tuple can return a bool directly, since the `keptProjections` are never used anyways.

**Review by bpintea — APPROVED** [review]
Left a more relevant note in a [comment](https://github.com/elastic/elasticsearch/pull/104958#discussion_r1476093868) and some smaller nits.
Otherwise, while not fundamental, this PR together with #104387 make ESQL writing really "liberating" -- very nice.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java [hunk 16]
```diff
@@ -148,23 +149,20 @@ private static void checkAggregate(LogicalPlan p, Set<Failure> failures) {
             // check aggregates
             agg.aggregates().forEach(e -> {
                 var exp = Alias.unwrap(e);
-                if (exp instanceof AggregateFunction af) {
+                Holder<Boolean> foundAgg = new Holder<>(false);
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:152]: True but that still requires a var and extra code to do the substraction and comparison...

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java (hunk 15 above)
Comment (costin) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:365]: It is:
```
                   if (tuple.v1().isEmpty()) {
                        var newAggs = tuple.v2();
                        var newGroups = replacePrunedAliasesUsedInGroupBy(a.groupings(), aggs, newAggs);
                        plan = new Aggregate(a.source(), a.child(), newGroups, newAggs);
                    }
```

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java [hunk 17]
```diff
@@ -1554,6 +1554,97 @@ public void testUnresolvedMvExpand() {
         assertThat(e.getMessage(), containsString("Unknown column [bar]"));
     }
 
+    public void testLiteralInAggregateNoGrouping() {
+        var e = expectThrows(VerificationException.class, () -> analyze("""
+             from test
+            |stats 1
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java:1560]: I prefer to have separate tests since in case of a failure it is to identify what case failed based on the method name:
`testLiteralBehindEvalInAggregateNoGrouping`.
Tests are one area where I prefer to be verbose and thus kind to my future self/developer who might investigate what's going on.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec [hunk 18]
```diff
@@ -1112,3 +1112,37 @@ STATS ck = COUNT(job_positions),
...
+;
+
+x:d       | a:d      | m:i
+99123.275 | 48248.55 |74999
+;
+
+nestedAggsWithGrouping#[skip:-8.12.99,reason:supported in 8.13+]
+FROM employees
+| STATS x = ROUND(AVG(salary % 3)) + MAX(emp_no), y = MIN(emp_no / 3) + 10 - MEDIAN(salary) by z = languages % 2
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:1127]: This rule acts as syntactic sugar in a way - some might prefer to use eval, others might not. I don't think there's a right or wrong way, rather a stylistic preference.

[the remark this answers is not in the document]
Comment (bpintea) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:1127]: The nice thing with this new flexibility is that we could detect if the user wants an agg or just an eval, we'd no longer need to have two distinctive commands. I think this is what Alex was suggesting -- it's an interesting idea.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java [hunk 19]
```diff
@@ -148,23 +149,20 @@ private static void checkAggregate(LogicalPlan p, Set<Failure> failures) {
...
-                                "expected an aggregate function or group but got ["
-                                    + exp.sourceText()
-                                    + "] of type ["
-                                    + exp.nodeName()
-                                    + "]"
-                            )
-                        );
-                    }
+                });
+                if (foundAgg.get() == false && Expressions.match(agg.groupings(), g -> Alias.unwrap(g).semanticEquals(exp)) == false) {
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:159]: Improved it to indicate if it's a reference to the actual group which is not allowed in the declaration:
"expected an aggregate function but got [x] referencing [first_name]"

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java (hunk 11 above)
Comment (costin) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:1265]: Added a test for it just in case.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java (hunk 7 above)
Comment (costin) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:1232]: That would be a nice little feature - raised https://github.com/elastic/elasticsearch/issues/105102 as a follow-up.

**Comment by costin** [comment]
> * I think there's bugs for the following cases: `stats max(l) by l=languages` (verification exception) and, more severely, `stats max(languages) + languages by l = languages` (NPE, `Cannot invoke \"org.elasticsearch.xpack.esql.planner.Layout$ChannelAndType.channel()\" because the return value of \"org.elasticsearch.xpack.esql.planner.Layout.get(org.elasticsearch.xpack.ql.expression.NameId)\" is null"`); the latter works fine without the alias.

I couldn't reproduce these two queries - either they are kicked off by the verifier (cannot use aggregate over grouping) or don't resolve (l cannot be resolved since it is used by the agg function but is declared afterwards).
Added them in the verifier test suite.

**Comment by costin** [comment]
> * I think there's bugs for the following cases: `stats max(l) by l=languages` (verification exception) and, more severely, `stats max(languages) + languages by l = languages` (NPE, `Cannot invoke \"org.elasticsearch.xpack.esql.planner.Layout$ChannelAndType.channel()\" because the return value of \"org.elasticsearch.xpack.esql.planner.Layout.get(org.elasticsearch.xpack.ql.expression.NameId)\" is null"`); the latter works fine without the alias.
> * This allows shenanigans like `stats max(languages) + languages by languages` (using a grouping in the expression), but not `stats languages + 1 by languages`; although that may be something for a follow-up, if we want to allow this.

There's a subtle bug when dealing with an aliased grouping which sometimes popped up because the validation allowed the grouping column in some queries.
I've raised #105172 as a follow-up and in the meantime improved the validator to not allow this scenario.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java [hunk 20]
```diff
@@ -2895,6 +2921,216 @@ public void testNestedMultiExpressionsInGroupingAndAggs() {
         assertThat(Expressions.names(agg.output()), contains("count(salary + 1)", "max(salary   +  23)", "languages   + 1", "emp_no %  3"));
     }
 
+    /**
+     * Expects
+     * Project[[x{r}#5]]
+     * \_Eval[[____x_AVG@9efc3cf3_SUM@daf9f221{r}#18 / ____x_AVG@9efc3cf3_COUNT@53cd08ed{r}#19 AS __x_AVG@9efc3cf3, __x_AVG@
+     * 9efc3cf3{r}#16 / 2[INTEGER] + __x_MAX@475d0e4d{r}#17 AS x]]
+     *   \_Limit[500[INTEGER]]
+     *     \_Aggregate[[],[SUM(salary{f}#11) AS ____x_AVG@9efc3cf3_SUM@daf9f221, COUNT(salary{f}#11) AS ____x_AVG@9efc3cf3_COUNT@53cd0
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:2930]: I've revisited the naming strategy to rely on a counter across the entire rule - hopefully this avoids clashes and improves readability.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 21]
```diff
@@ -205,7 +204,7 @@ protected LogicalPlan rule(Aggregate aggregate) {
                             var attr = aggFuncToAttr.get(af);
                             // the agg doesn't exist in the Aggregate, create an alias for it and save its attribute
                             if (attr == null) {
-                                var temporaryName = temporaryName(agg, af);
+                                var temporaryName = temporaryName(af, agg, counter[0]++);
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:207]: Changed the strategy to indicate the inner and outer expression as that can differ across rules - in some an aggregation is an inner expressions while in others is the outer one.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 22]
```diff
@@ -239,15 +238,31 @@ protected LogicalPlan rule(Aggregate aggregate) {
...
             return plan;
         }
 
-        static String temporaryName(NamedExpression agg, AggregateFunction af) {
-            return "__" + agg.name() + "_" + af.functionName() + "@" + Integer.toHexString(af.hashCode());
+        static String temporaryName(Expression inner, Expression outer, int suffix) {
+            String in = toString(inner);
+            String out = toString(outer);
+            return "$$" + in + "$" + out + "$" + suffix;
+        }
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:252]: Opted for $$a$ instead of _ as that was used to replace spaces.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java [hunk 23]
```diff
@@ -2895,6 +2921,267 @@ public void testNestedMultiExpressionsInGroupingAndAggs() {
...
+     * >$MAX$1{r}#45 AS g]]
+     * \_Eval[[$$$$avg(salary)_+_m>$AVG$0$SUM$0{r}#48 / $$max(salary)_+_3>$COUNT$2{r}#46 AS $$avg(salary)_+_m>$AVG$0, $$avg(
+     * salary)_+_m>$AVG$0{r}#44 + $$avg(salary)_+_m>$MAX$1{r}#45 AS a, $$avg(salary)_+_m>$MAX$1{r}#45 + 3[INTEGER] +
+     * 3.141592653589793[DOUBLE] + $$max(salary)_+_3>$COUNT$2{r}#46 AS b]]
+     *   \_Limit[500[INTEGER]]
+     *     \_Aggregate[[w{r}#28],[SUM(salary{f}#39) AS $$$$avg(salary)_+_m>$AVG$0$SUM$0, MAX(salary{f}#39) AS $$avg(salary)_+_m>$MAX$1
+     * , COUNT(salary{f}#39) AS $$max(salary)_+_3>$COUNT$2, MIN(salary{f}#39) AS $$count(salary)_->$MIN$3]]
+     *       \_Eval[[languages{f}#37 % 2[INTEGER] AS w]]
+     *         \_EsRelation[test][_meta_field{f}#40, emp_no{f}#34, first_name{f}#35, ..]
+     */
```
Comment (costin) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:3091]: An example of the new naming strategy.
